test/config-luatest/set_names_test.lua

```diff
@@ -0,0 +1,132 @@
+    g.replica_set:wait_for_fullmesh()
+    for _, replica in ipairs(g.replica_set.servers) do
+        replica:exec(function()
+            box.snapshot()
+        end)
+    end
+
+    g.replica_set:stop()
+    return instance_1.workdir, instance_2.workdir
+end
+
+g.before_all(function(g)
+    treegen.init(g)
+    local uuids = {
+        ['replicaset-001'] = uuid.str(),
+        ['instance-001'] = uuid.str(),
+        ['instance-002'] = uuid.str(),
+    }
+
+    local workdir_1, workdir_2 = initialize_xlogs(g, uuids)
+    local config = {
+        credentials = {
+            users = {
+                guest = {
+                    roles = {'super'},
+                },
+            },
+        },
+
+        iproto = {
+            listen = 'unix/:./{{ instance_name }}.iproto',
+        },
+
+        groups = {
+            ['group-001'] = {
+                replicasets = {
+                    ['replicaset-001'] = {
+                        database = {
+                            replicaset_uuid = uuids['replicaset-001']
+                        },
+                        instances = {
+                            ['instance-001'] = {
+                                snapshot = { dir = workdir_1, },
+                                wal = { dir = workdir_1, },
+                                database = {
+                                    instance_uuid = uuids['instance-001'],
+                                    mode = 'rw',
+                                },
+                            },
+                            ['instance-002'] = {
+                                snapshot = { dir = workdir_2, },
+                                wal = { dir = workdir_2, },
+                                database = {
+                                    instance_uuid = uuids['instance-002']
+                                },
+                            },
+                        },
+                    },
+                },
+            },
+        },
+    }
+
+    local cfg = yaml.encode(config)
+    local dir = treegen.prepare_directory(g, {}, {})
+    local config_file = treegen.write_script(dir, 'cfg.yaml', cfg)
+    local opts = {config_file = config_file, chdir = dir}
+    g.instance_1 = server:new(fun.chain(opts, {alias = 'instance-001'}):tomap())
+    g.instance_2 = server:new(fun.chain(opts, {alias = 'instance-002'}):tomap())
+
+    g.instance_1:start({wait_until_ready = false})
+    g.instance_2:start({wait_until_ready = false})
+    g.instance_1:wait_until_ready()
+    g.instance_2:wait_until_ready()
+end)
+
+g.after_all(function(g)
+    g.replica_set:drop()
+    g.instance_1:drop()
+    g.instance_2:drop()
+    treegen.clean(g)
+end)
+
+local function check_names(rs_name, name)
+    t.helpers.retrying({}, function()
+        local info = box.info
+        t.assert_equals(info.replicaset.name, rs_name)
+        t.assert_equals(info.name, name)
+
+        t.assert_equals(box.cfg.replicaset_name, rs_name)
```

Comment:
1 out of 100 runs of this test fails with the following log:

<details>
<summary>Logs</summary>

```
[010] instance-002 | 2023-10-24 02:05:54.867 [505343] main/103/interactive I> update replication_synchro_quorum = 2
[010] instance-002 | 2023-10-24 02:05:54.867 [505343] main/111/applier/unix/:./instance-001.iproto I> assigned name instance-001 to replica c4a61f72-7a0f-471e-a918-7f599704f1ec
[010] instance-002 | 2023-10-24 02:05:54.867 [505343] main/111/applier/unix/:./instance-001.iproto I> assigned name instance-002 to replica 6b62463b-80d6-4ac1-8bff-4d5e0c34df15
[010] instance-002 | 2023-10-24 02:05:54.868 [505343] main/106/checkpoint_daemon I> scheduled next checkpoint for Tue Oct 24 03:51:04 2023
[010] instance-002 | 2023-10-24 02:05:54.869 [505343] main/124/console/unix/:./var/run/instance-002/tarantool.control/socket I> started
[010] instance-002 | 2023-10-24 02:05:54.869 [505343] main/103/interactive/box.load_cfg I> set 'instance_name' configuration option to "instance-002"
[010] instance-002 | 2023-10-24 02:05:54.869 [505343] main/103/interactive/tarantool.config log.lua:130 W> box_cfg.apply: name replicaset-001 for 4c0a59dd-49ab-4dae-80f9-e18767f8f0a5 is missing from the snapshot.
[010] instance-002 | 2023-10-24 02:05:54.870 [505343] main/123/lua/tarantool.config log.lua:130 W> credentials.apply: Tarantool is in Read Only mode, so credentials will be set up in the background when it is switched to Read Write mode
[010] instance-002 | 2023-10-24 02:05:54.870 [505343] main I> entering the event loop
[010] instance-002 | 2023-10-24 02:05:54.870 [505343] main I> replicaset name: replicaset-001
[010] not ok 1	set-names-automatically.test_names_are_set
[010] #   ...ing/tnt/tarantool/test/config-luatest/set_names_test.lua:119: expected: "replicaset-001"
[010] #   actual: nil
[010] #   stack traceback:
[010] #   	...ing/tnt/tarantool/test/config-luatest/set_names_test.lua:131: in function 'set-names-automatically.test_names_are_set'
[010] #   	...
[010] #   	[C]: in function 'xpcall'
```

</details>

`instance-002` gets all instance names, make an alert, that `replicaset_name` is missing (which means, on_commit trigger on schema for setting REPLICASET_NAME is not yet executed), sets on_replace trigger and gets replicaset_name. But  out on_replace trigger is never executed (I have print right inside it).

It seems, that tarantool yields between executing on_replace triggers and on_commit. When we're right inside this gap, our on_replace is not executed, but the name is already set. Strange, that this happens only with `replicaset_name`.  `set_names_reload_test` is subject to the same error(

Reply:
> It seems, that tarantool yields between executing on_replace triggers and on_commit. When we're right inside this gap, our on_replace is not executed, but the name is already set. 

Yep, that's possible. This wouldn't be a problem if you set `on_replace` triggers right from `on_schema_init`.
I remember you writing that this was impossible, but I don't remember why and I can't find the place mentioning it now.

Reply:
> This wouldn't be a problem if you set on_replace triggers right from on_schema_init.

Great idea! This fixes the flakiness